### ISSUE TITLE
CTRL-C does not kill cross

### ISSUE THREAD

**Issue by TheButlah** [cross-rs/cross#1041]
### Checklist

- [X] I've looked through the [issues and pull requests](https://github.com/cross-rs/cross/issues?q=) for similar reports

### Describe your issue

Hitting CTRL-C on my macbook fails to kill a `cross build`. Instead, I must kill this process with `kill <PID>` or `docker kill <CONTAINER>`

### What target(s) are you cross-compiling for?

x86_64-unknown-linux-gnu

### Which operating system is the host (e.g computer cross is on) running?

- [X] macOS
- [ ] Windows
- [ ] Linux / BSD
- [ ] other OS (specify in description)

### What architecture is the host?

- [ ] x86_64 / AMD64
- [ ] arm32
- [X] arm64 (including Mac M1)

### What container engine is cross using?

- [X] docker
- [ ] podman
- [ ] other container engine (specify in description)

### cross version

cross 0.2.4,  cargo 1.66.0-nightly (73ba3f35e 2022-09-18)

### Example

```sh
cross build --target x86_64-unknown-linux-gnu 
```

### Additional information / notes

Everything works correctly when the `aarch64-apple-darwin` (host) target is used instead

**Comment by Alexhuszagh** [comment]
The host target `aarch64-apple-darwin` falls back to the host cargo, and is not run inside a container, so this wouldn't be cross's domain. The `Ctrl+C` event not being captured might be my fault....
https://github.com/cross-rs/cross/pull/964

**Comment by Alexhuszagh** [comment]
So this still fails if a process is spawned in the background. A quick test is:

```bash
$ cross build --target armv7-unknown-linux-gnueabi &
$ jobs
[1]+  Stopped                 cross build --target armv7-unknown-linux-gnueabi
[1]+  Stopped                 cross build --target armv7-unknown-linux-gnueabi
$ fg
cross build --target armv7-unknown-linux-gnueabi
    Finished dev [unoptimized + debuginfo] target(s) in 0.16s
$ jobs
```

There might be a better way than adding `-i` for an interactive prompt so we can both kill the docker container if a signal is sent to `cross` and propagate that to the container. The termination handlers we added might be useful here.

**Comment by sadilekivan** [comment]
I still encounter this problem when my cross runner opens an ssh session to run remotely. It's really annoying to restart the terminal everytime. I tried to google around on how to pass ctrl+c to the underlying container but cannot figure this out. Anything obvious I could do to pass `SIGTERM`

`./remote-runner`
```bash
#!/bin/bash

local_path=$1
shift
name=$(basename $local_path)
remote_path="/tmp/$name"
ssh_flags="-o \"UserKnownHostsFile=/dev/null\" -o \"StrictHostKeyChecking no\" -o \"LogLevel ERROR\""

transfer_command="rsync $local_path $CARGO_TARGET_HOSTNAME:$remote_path -e 'ssh $ssh_flags'"
execute_command="ssh -t $ssh_flags $CARGO_TARGET_HOSTNAME 'RUST_BACKTRACE=$RUST_BACKTRACE RUST_LOG=$RUST_LOG $remote_path $@'"

eval $transfer_command && \
eval $execute_command
```

`Cross.toml`
```toml
[build]
default-target = "aarch64-unknown-linux-gnu"

[target.aarch64-unknown-linux-gnu]
dockerfile = "./dockerfile"

[target.aarch64-unknown-linux-gnu.env]
passthrough = [
    "CARGO_TARGET_AARCH64_UNKNOWN_LINUX_GNU_RUNNER=./remote-runner",
    "CARGO_TARGET_HOSTNAME=root@10.0.0.1",
    "RUST_BACKTRACE=1",
    "RUST_LOG=trace"
]
```

If I try to use my runner outside cross/docker container it captures `ctrl+c` fine

**Comment by mcandre** [comment]
This misbehavior is happening again with cross v0.2.5.

If the user sends SIGINT (e.g. Ctrl+C on macOS) to cross *just before* cross emits any output, then cross enters a bad program state and stalls indefinitely. The Docker processes continue to spin their wheels as well.

Recommendations for a workaround: Use a portable signal handler like the ctrlc crate; use an AtomicBool to communicate whether the user has requested early termination or not; avoid spawning child processes in the event of early termination; and ensure that all child processes have a wait method applied.

https://crates.io/crates/ctrlc

**Comment by Emilgardis** [comment]
the pr for the fix is not included in 0.2.5, it's expected that this is still an issue in 0.2.5

**Comment by mcandre** [comment]
Update:

cross now exits mostly cleanly when I send SIGINT / Ctrl+C to my wrapping CLI app. No explicit signal handler involved.

However, two caveats remain.

1. I have to setup the stdout and stderr cross process streams as piped, rather than the default inherited configuration.
2. cross still leaves behind Docker containers when cross is killed.
